server/sonar-server/src/test/java/org/sonar/server/rule/RegisterRulesTest.java

```diff
@@ -600,6 +603,64 @@ public void rules_that_remove_deprecated_key_must_remove_records() {
     assertThat(deprecatedRuleKeys).hasSize(0);
   }
 
+  @Test
+  public void declaring_two_rules_with_same_deprecated_RuleKey_should_throw_ISE() {
+    expectedException.expect(IllegalStateException.class);
+    expectedException.expectMessage("The following deprecated rule keys are declared at least twice [fake:old]");
+
+    execute(context -> {
+      NewRepository repo = context.createRepository("fake", "java");
+      repo.createRule("newKey1")
+        .setName("One")
+        .setHtmlDescription("Description of One")
+        .setSeverity(BLOCKER)
+        .setInternalKey("config1")
+        .setTags("tag1", "tag2", "tag3")
+        .setType(RuleType.CODE_SMELL)
+        .addDeprecatedRuleKey("fake", "old")
+        .setStatus(RuleStatus.BETA);
+      repo.createRule("newKey2")
+        .setName("One")
+        .setHtmlDescription("Description of One")
+        .setSeverity(BLOCKER)
+        .setInternalKey("config1")
+        .setTags("tag1", "tag2", "tag3")
+        .setType(RuleType.CODE_SMELL)
+        .addDeprecatedRuleKey("fake", "old")
+        .setStatus(RuleStatus.BETA);
+      repo.done();
+    });
+  }
+
+  @Test
+  public void declaring_a_rule_with_a_deprecated_RuleKey_still_used_should_throw_ISE() {
```

Comment:
a unit test is missing for two rules declaring the same key. The new code will detect it too, if I'm not mistaken.

Reply:
I've created one (with the previous ` List<RuleKey> definedRuleKeys = allRules.stream()` for detecting duplicates but I've found that `RulesDefinition.NewRepositoryImpl#createRule` is already detecting duplicates.
I'll readd a test on this.